src/abstractdataframe/selection.jl

```diff
@@ -221,153 +221,6 @@ const TRANSFORMATION_COMMON_RULES =
-                if !all(x -> x isa Pair && last(x) == dst, second)
-                    throw(ArgumentError("when using broadcasted column selector " *
-                                        "it must have a constant value"))
-                end
-                need_broadcast = true
-                new_names = names(df, dst.sel)
-                if !(length(new_names) == size(p, 1) || size(p, 1) == 1)
-                    throw(ArgumentError("broadcasted dimension does not match the " *
-                                        "number of selected columns"))
-                end
-                new_dst = new_names
-                new_second = first.(second) .=> new_dst
-            else
-                new_second = second
-            end
-        else
-            new_second = second
-        end
-    end
-
-    if need_broadcast
-        new_p = new_src .=> new_second
-        return isempty(new_p) ? [] : new_p
-    else
-        return p
-    end
-end
-
-# add a method to funname defined in other/utils.jl
-funname(row::ByRow) = funname(row.fun)
-
-make_pair_concrete(@nospecialize(x::Pair)) =
-    make_pair_concrete(x.first) => make_pair_concrete(x.second)
-make_pair_concrete(@nospecialize(x)) = x
-
-normalize_selection(idx::AbstractIndex, @nospecialize(sel), renamecols::Bool) =
-    try
-        idx[sel]
-    catch e
-        if e isa MethodError && e.f === getindex && e.args === (idx, sel)
-            throw(ArgumentError("Unrecognized column selector $sel in AsTable constructor"))
-        else
-            rethrow(e)
-        end
-    end
-
-normalize_selection(idx::AbstractIndex, @nospecialize(sel::Base.Callable), renamecols::Bool) = sel
-normalize_selection(idx::AbstractIndex, sel::Colon, renamecols::Bool) = idx[:]
-
-normalize_selection(idx::AbstractIndex, sel::Pair{typeof(nrow), Symbol},
-                    renamecols::Bool) =
-    length(idx) == 0 ? (Int[] => (() -> 0) => last(sel)) : (1 => length => last(sel))
-normalize_selection(idx::AbstractIndex, sel::Pair{typeof(nrow), <:AbstractString},
-                    renamecols::Bool) =
-    normalize_selection(idx, first(sel) => Symbol(last(sel)), renamecols)
-normalize_selection(idx::AbstractIndex, sel::typeof(nrow), renamecols::Bool) =
-    normalize_selection(idx, nrow => :nrow, renamecols)
-
-normalize_selection(idx::AbstractIndex, sel::Pair{typeof(eachindex), Symbol},
-                    renamecols::Bool) =
-    length(idx) == 0 ? (Int[] => (() -> Base.OneTo(0)) => last(sel)) : (1 => eachindex => last(sel))
-normalize_selection(idx::AbstractIndex, sel::Pair{typeof(eachindex), <:AbstractString},
-                    renamecols::Bool) =
-    normalize_selection(idx, first(sel) => Symbol(last(sel)), renamecols)
-normalize_selection(idx::AbstractIndex, sel::typeof(eachindex), renamecols::Bool) =
-    normalize_selection(idx, eachindex => :eachindex, renamecols)
-
 normalize_selection(idx::AbstractIndex, sel::Pair{typeof(groupindices), Symbol},
```

Comment:
Shouldn't these methods be moved to the same place as others? Code will become quite complex if methods are split across files/packages.

Reply:
The `normalize_selection`s that were leftover have a dependency on `groupindices` and `proprow`, which I didn't move out of DataFrames.
It's not difficult to move them out, so I suppose we should do it for the sake of completeness?

Reply:
I think we should move everything for completeness, i.e. we should make sure that any operation specification syntax works the same on DataFrames.jl and DTables.jl. Even if in DTables.jl you do not provide these functionalities it should throw an informative error (i.e. that operation is not supported yet instead of producing some garbage or incorrect result).

Having said that I want to ask the following question that I wanted to ask for some time now (but hesitated as I know you spent a lot of time working on this PR):

> Let us summarize the pros and cons of splitting the interface from DataFrames.jl

I understand there are two options (I list their pros):

Split the module and make it DTables.jl dependency:
* faster loading time (DataFrames.jl load time is 1 second); the question is if this is indeed an important factor? (i.e. most likely e.g. starting Julia cluster anyway takes much longer)
* less indirect dependencies (this could be relevant, but we make sure to keep the number of dependencies of DataFrames.jl low and only allow well maintained ones)
* less frequent updates of dependency (assuming that operation specification syntax is updated less frequently than the whole package; but I think that the reality is that "something" will change here with every release minor release of DataFrames.jl anyway)

Make DTables.jl depend on DataFrames.jl:
* you do not have to export DataFrames.jl if you do not want to (you can just `import` it internally)
* you will be in sync with DataFrames.jl automatically (i.e. there will be no issue - like the issue that prompted me to write this comment that API is only partially supported by the interface package)
* you can use DataFrames.jl fallback methods without a problem (we will not have to go into a long discussion what methods should go into DataAPI.jl interface and what interface they should have - if you see that the user uses DataFrames.jl as a backend for computations you can easily use specialized methods from DataFrames.jl to do partial steps in DTables.jl)
* most of your users will use DataFrames.jl most likely anyway, so it is better to make sure we are in sync by design.
* whatever dependencies DataFrames.jl takes you can be sure that we track them with our tests (and I assure you that we have caught dozens of errors in external packages this way; the same is with user base - they constantly check if what we deploy actually works) - so, this puts a less burden on DTables.jl with making sure that all packages you depend on keep working correctly (except packages that you want to depend on and DataFrames.jl does not depend on)

In summary:
* taking whole DataFrames.jl does not seem to me as a huge limitation for DTables.jl development (you do not have to export anything; and can decide to leverage)
* you will be able to call methods from DataFrames.jl without a problem if you detect that underlying table type that the user uses is DataFrames.jl type (this does not restrict you from handling a general case using other code path).
* users will get a clear message that "newbie friendly" combination is DTables.jl + DataFrames.jl (so they do not have to think what to choose); while experts still are free to choose whatever they wish.

What would be needed to be done if we decided not to split the package from DataFrames.jl is to document the API you want to use in DTables.jl as public, so that you can safely rely on it (and that is why the work you do now is anyway useful, because we learn what actually should be a part of this API).

Please let me know what you think.